compiler/src/main/kotlin/walkers/symbol.kt

```diff
@@ -0,0 +1,60 @@
+package dk.aau.cs.d409f19.cellumata.walkers
+
+import dk.aau.cs.d409f19.cellumata.ast.*
+import java.lang.Exception
+
+class SymbolException(val ident: String) : Exception("\"$ident\" was used before it was declared")
+
+class ScopeCheckVisitor(val symbolTable: SymbolTable) : BaseASTVisitor() {
+    override fun visit(node: RootNode) {
+        // Extract all states, functions, and neighbourhoods names first
+        node.body.filter { it is StateDecl }.forEach { symbolTable.insertSymbol((it as StateDecl).ident, it) }
+        node.body.filter { it is FuncDecl }.forEach { symbolTable.insertSymbol((it as FuncDecl).ident, it) }
+        node.body.filter { it is NeighbourhoodDecl }.forEach { symbolTable.insertSymbol((it as NeighbourhoodDecl).ident, it) }
+
+        // Visit constant declaration first
+        node.body.filter { it is ConstDecl }.forEach { visit(it) }
+
+        // Visit the other declarations
+        node.body.filter { it !is ConstDecl }.forEach { visit(it) }
+    }
+
+    override fun visit(node: VarExpr) {
+        if (symbolTable.getSymbol(node.ident) == null) {
+            throw SymbolException(node.ident)
+        }
+    }
+
+    override fun visit(node: ConstDecl) {
+        super.visit(node)
+        symbolTable.insertSymbol(node.ident, node)
+    }
+
+    override fun visit(node: StateDecl) {
+        symbolTable.createScope()
+        super.visit(node)
+        symbolTable.closeScope()
+    }
+
+    override fun visit(node: FuncDecl) {
+        symbolTable.createScope()
+        node.args.forEach { symbolTable.insertSymbol(it.ident, it) }
+
+        super.visit(node)
+        symbolTable.closeScope()
+    }
+
+    override fun visit(node: AssignStmt) {
+        super.visit(node)
+
+        if (node.ctx.STMT_LET() != null) { // Check if this is a variable declaration, and not just an assignment
+            symbolTable.insertSymbol(node.ident, node)
+        }
+    }
+
+    override fun visit(node: ConditionalBlock) {
+        symbolTable.createScope()
+        super.visit(node)
+        symbolTable.closeScope()
+    }
+}
```

Comment:
You have VarExpr where symbol declaration is checked, but FuncExpr, StateExpr, NeighbourhoodExpr (or whatever their names are) are not checked if the name is already declared.

Reply:
State names and neighbourhood names are currently matched by VarExpr, so maybe we should rename VarExpr to something more generic.

Reply:
e662dc1 implements code that checks functions are declared before they are used.

Reply:
I suggest `Identifier` for named expressions. Or `NamedExpression`.

Reply:
I will resolve and approve after rename.

Reply:
I have renamed `VarExpr` to `NamedExpr` in 8247d7f